pensieve/export_json.py

```diff
@@ -0,0 +1,114 @@
+from google.cloud import bigquery
+from google.cloud import storage
+import logging
+import smart_open
+from typing import Dict
+
+logging.getLogger(__name__)
+
+
+def _get_statistics_tables(client: bigquery.Client, bq_dataset: str) -> Dict[str, str]:
+    """Returns statistics table names and their last modified timestamp."""
+    job = client.query(
+        f"""
+        SELECT table_id, TIMESTAMP_MILLIS(last_modified_time) as last_modified
+        FROM {bq_dataset}.__TABLES__
+        WHERE table_id LIKE 'statistics_%'
+    """
+    )
+
+    result = job.result()
+    return {row.table_id: row.last_modified for row in result}
+
+
+def _get_gcs_blobs(storage_client: storage.Client, bucket: str) -> Dict[str, str]:
+    """Return all blobs in the GCS location with their last modified timestamp."""
+    blobs = storage_client.list_blobs(bucket)
+
+    return {blob.name.replace(".json", ""): blob.updated for blob in blobs}
+
+
+def _export_table(
+    client: bigquery.Client,
+    project_id: str,
+    dataset_id: str,
+    table: str,
+    bucket: str,
+    storage_client: storage.Client,
+):
+    """Export a single table or view to GCS as JSON."""
+    # since views cannot get exported directly, write data into a temporary table
+    job = client.query(
+        f"""
+        SELECT *
+        FROM {dataset_id}.{table}
+    """
+    )
+
+    job.result()
+
+    # get the temporary table results are written to
+    tmp_table = job._properties["configuration"]["query"]["destinationTable"]
+
+    destination_uri = f"gs://{bucket}/{table}.ndjson"
+    dataset_ref = bigquery.DatasetReference(project_id, tmp_table["datasetId"])
+    table_ref = dataset_ref.table(tmp_table["tableId"])
+
+    logging.info(f"Export table {table} to {destination_uri}")
+
+    job_config = bigquery.ExtractJobConfig()
+    job_config.destination_format = "NEWLINE_DELIMITED_JSON"
+    extract_job = client.extract_table(
+        table_ref, destination_uri, location="US", job_config=job_config
+    )
+    extract_job.result()
+
+    # convert ndjson to json
+    _convert_ndjson_to_json(bucket, table, storage_client)
+
+
+def _convert_ndjson_to_json(bucket_name: str, table: str, storage_client: storage.Client):
+    """Converts the provided ndjson file on GCS to json."""
+    ndjson_blob_path = f"gs://{bucket_name}/{table}.ndjson"
+    json_blob_path = f"gs://{bucket_name}/{table}.json"
+
+    logging.info(f"Convert {ndjson_blob_path} to {json_blob_path}")
+
+    # stream from GCS
```

Comment:
is this fast? these files should be small enough that we'd never care about streaming them but if smart_open has good buffering behavior then this seems fine.

Reply:
The files are pretty small (a couple of kilobytes), not sure if they potentially get larger in the future. In any case, smart_open seems pretty fast and the whole export process for a table takes only a few seconds.